### DISCUSSION TITLE
typing patterns for cached data results

### DISCUSSION THREAD

**Discussion by jvanasco** [sqlalchemy/dogpile.cache#256] [Usage Questions] [answered]
The returned values of `CacheRegion.get(key)` and `CacheRegion.get_or_create(key, generator_function)` are typed as `CachedValue` and `NoValue`.

Trying to type the `CachedValue` in one of my projects is rather difficult because I utilize a proxy pattern.  A stripped down description is the following:

    CACHE_REGIONS: Dict[str, CacheRegion]: {"region1": region1CacheInstance}

    class CacheProxyA():
         _cache_region = "region1"
        def guid_to_id(self, guid:str): 
           key = "guid2id:%s" % guid
           return CACHE_REGIONS[self._cache_region].get(key)

The issue is that I need to type the return value of the function as `-> Union[NoValue, int]`, as various filters will deserialize/decode a cache hit into the int.

If I type the return value, so calling functions are correct, I get a type mismatch error:

    error: Incompatible return value type (got "CachedValue | NoValue", expected "int | NoValue") 

If I don't type the return value, no issue there, but I lose the benefits of typing.

Does anyone have any ideas on possible solutions?

I know there is no perfect solution, as a lot of this is handled through configured dogpile filters/proxies that the typing system is oblivious to.

At the very least, I'd like to match the typing of the output (`CacheRegion.get`) with the typing of the input (the output of the function passed to `CacheRegion.get_or_create`).

**Comment by zzzeek** [comment, marked as answer]
we are working ont his over at https://gerrit.sqlalchemy.org/c/sqlalchemy/dogpile.cache/+/5270

**Reply by jvanasco** [reply]
Thanks, Mike. I'll follow that and start reading up on the changes.
